src/main/java/com/salesforce/dataloader/security/EncryptionAesUtil.java

```diff
@@ -0,0 +1,255 @@
+            if (detectedOS == OSType.MacOS || detectedOS == OSType.Linux) {
+                // set all reading to false
+                customDir.setReadable(false, false);
+                // only owner can read
+                customDir.setReadable(true, true);
+            }
+            LOGGER.info(customDir + " was created");
+
+        } else {
+            LOGGER.info(customDir + " was not created");
+            throw new RuntimeException("Cannot create directory:" + path);
+        }
+        return Paths.get(path, DEFAULT_KEYFILE_NAME).toString();
+    }
+
+    // If given path not existing, create one.
+    public String createKeyFileIfNotExisting(String filePath) throws GeneralSecurityException {
+        if (filePath == null || filePath.isEmpty() || !Files.exists(Paths.get(filePath))) {
+            // no valid file path provided, check if default one exists
+            if (filePath == null || filePath.isEmpty()) {
+                filePath = createUserProfileKeyName();
+            }
+            if (!Files.exists(Paths.get(filePath))) {
+                byte[] key = generateEncryptionKey();
+                try (FileOutputStream fos = new FileOutputStream(filePath)) {
+                    fos.write(key);
+                } catch (IOException io) {
+                    throw new GeneralSecurityException("Failed to open file:" + filePath, io);
+                }
+                // Windows platform is already readable only to owner.
+                if (detectedOS == OSType.MacOS || detectedOS == OSType.Linux) {
+                    File file = new File(filePath);
+                    // set all reading to false
+                    file.setReadable(false, false);
+                    // set Owner reading to true
+                    file.setReadable(true, true);
+                }
+            }
+        }
+        setCipherKeyFromFilePath(filePath);
+        return filePath;
+    }
+
+    public void resetCipherKey() {
+        cipherKey = null;
+    }
+
+    public void setCipherKeyFromFilePath(String filePath) throws GeneralSecurityException {
+        byte[] data = new byte[1024];
+        try (FileInputStream fis = new FileInputStream(filePath)) {
+            int size = fis.read(data);
+            if (size < ENCRYPTION_KEY_LENGTH_IN_BYTES)
+                throw new GeneralSecurityException("Keyfile content is too short:" + filePath);
+            else {
+                cipherKey = new byte[ENCRYPTION_KEY_LENGTH_IN_BYTES];
+                System.arraycopy(data, 0, cipherKey, 0, ENCRYPTION_KEY_LENGTH_IN_BYTES);
+            }
+        } catch (IOException io) {
+            throw new GeneralSecurityException("Failed to open file: " + filePath, io);
+        }
+    }
+
+    private void ensureKeyIsSet() throws IOException, GeneralSecurityException {
+        if (cipherKey != null) {
+            return;
+        }
+        // if no key was set up already, create one
+        createKeyFileIfNotExisting(null);
```

Comment:
This function will set the `cipherKey` to the contents of the keyfile whether or not the keyfile was already existing, unless the keyfile is too short (in which case it won't replace it with a valid one, but will at least throw an exception). However, that series of checks and outcomes is not obvious from the function names (or comments, which are mostly missing); it might be better to make them more explicit